# Контрольная работа 1
Колченко Ксения

In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
import statsmodels.api as sm

## Задача 1

Для датасета 'wage2' оценим регрессию **log(wage) на IQ, south**

In [4]:
df_wage = pd.read_csv('C:\Users\stud.LIMM\Downloads\Econometrica-master\python-notebooks\data-csv\wage2.csv')
df_wage.head()

wage  hours   IQ  KWW  educ  exper  tenure  age  married  black  south  \
0   769     40   93   35    12     11       2   31        1      0      0   
1   808     50  119   41    18     11      16   37        1      0      0   
2   825     40  108   46    14     11       9   33        1      0      0   
3   650     40   96   32    12     13       7   32        1      0      0   
4   562     40   74   27    11     14       5   34        1      0      0   

   urban  sibs  brthord  meduc  feduc     lwage  
0      1     1      2.0    8.0    8.0  6.645091  
1      1     1      NaN   14.0   14.0  6.694562  
2      1     1      2.0   14.0   14.0  6.715384  
3      1     4      3.0   12.0   12.0  6.476973  
4      1    10      6.0    6.0   11.0  6.331502

Найдем коэффиценты подогнанной модели

In [30]:
# специфицируем модель через формулу
wage_eq1 = smf.ols(formula='np.log(wage)~IQ+south', data=df_wage).fit()
# Коэффициенты модели с округлением
wage_eq1.params.round(2)

Intercept    6.01
IQ           0.01
south       -0.12
dtype: float64

## Задача 2

Для датасета 'wage2' оценим регрессию **log(wage) на age, south, urban, married**

In [10]:
# специфицируем модель через формулу
wage_eq2 = smf.ols(formula='np.log(wage)~age+south+urban+married', data=df_wage).fit()
# Коэффициенты модели с округлением
wage_eq2.params.round(2)

Intercept    5.89
age          0.02
south       -0.15
urban        0.18
married      0.20
dtype: float64

Вычислим $R^2$ и $R^2_{adj}$

In [12]:
R2 = wage_eq2.rsquared
R2_adj = wage_eq2.rsquared_adj
print("R2 =", R2.round(3))
print("R2_adj =", R2_adj.round(3))

R2 = 0.118
R2_adj = 0.114


## Задача 3

Для датасета 'Labour' оценим регрессию **log(output) на log(labour) & log(wage)**

In [17]:
df_labour = pd.read_csv('C:\Users\stud.LIMM\Downloads\Econometrica-master\python-notebooks\data-csv\Labour.csv')
# Подгонка модели
labour_eq1 = smf.ols(formula='np.log(output)~np.log(labour)+np.log(wage)', data=df_labour).fit()
labour_eq1.params.round(2)

Intercept        -5.97
np.log(labour)    0.85
np.log(wage)      1.07
dtype: float64

Вычислим TSS, ESS, RSS

In [18]:
TSS = labour_eq1.centered_tss
ESS = labour_eq1.ess
RSS = labour_eq1.ssr
print("TSS =", TSS.round(2))
print("ESS =", ESS.round(2))
print("RSS =", RSS.round(2))

TSS = 797.67
ESS = 692.45
RSS = 105.22


## Задача 4

Для датасета 'Labour' оценим регрессию **log(output) на log(capital), log(labour), log(wage)**

In [25]:
# Подгонка модели
labour_eq2 = smf.ols(formula='np.log(output)~np.log(capital)+np.log(labour)+np.log(wage)', data=df_labour).fit()
labour_eq2.params.round(2)

Intercept         -5.01
np.log(capital)    0.15
np.log(labour)     0.72
np.log(wage)       0.92
dtype: float64

Для наблюдения с номером 464 найдем фактическое значение зависимой переменной, предсказанное значение и остаток

In [26]:
# Фактичечкое
np.log(df_labour["output"][466]).round(3)

1.271

In [27]:
# Предсказанное
labour_eq2.fittedvalues[466].round(3)

1.342

In [28]:
# Остаток
labour_eq2.resid[466].round(3)

-0.07

## Задача 5

Для набора данных 'Labour' рассмотрим линейную регрессию **log(output) на log(capital), labour, wage**

Оценим регрессию на данных

Спецификация: $\log(output)=\beta_0+\beta_1\log(capital)+\beta_2labour+\beta_3wage+u$

In [31]:
# Подгонка модели
labour_eq3 = smf.ols(formula='np.log(output)~np.log(capital)+labour+wage', data=df_labour).fit()
labour_eq3.params.round(2)

Intercept          0.58
np.log(capital)    0.44
labour             0.00
wage               0.02
dtype: float64

Для наблюдений с номерами 2, 11, 24, 101, 349, 501 найдем фактическое значение зависимой переменной, предсказанное значение и остатки.

In [36]:
ind = np.array([3, 20, 56, 112, 351, 501])-1
df = pd.DataFrame({'i': ind + 1, 'Факт': np.log(df_labour["output"][ind]), 'Fitted': labour_eq3.predict()[ind], 'Остаток': np.log(df_labour["output"][ind]) - labour_eq3.predict()[ind]})
df

i      Факт    Fitted   Остаток
2      3  3.359733  2.861427  0.498307
19    20  0.748116  0.630456  0.117660
55    56  5.662240  6.553122 -0.890881
111  112  3.723433  3.358122  0.365311
350  351  0.748819  0.730315  0.018504
500  501  1.297354  2.281278 -0.983925

Вычислим TSS, ESS, RSS

In [39]:
TSS = labour_eq3.centered_tss
ESS = labour_eq3.ess
RSS = labour_eq3.ssr
print("TSS =", TSS.round(3))
print("ESS =", ESS.round(3))
print("RSS =", RSS.round(3))

TSS = 797.667
ESS = 522.54
RSS = 275.128


Вычислим $R^2$ и $R^2_{adj}$

Вычислим стандартрую ошибку регрессии

In [42]:
n = len(df_labour["output"])
k = 3
s = np.sqrt(RSS / (n - k - 1))
print("s =", s.round(4))

s = 0.6978


Спецификация: $\log(output)=0.58+0.44*\log(capital)+0*labour+0.02*wage+u$

### Интерпретация модели
1) При увеличении капитала на 1%, выпуск увеличивается на 0.44%
2) При увеличении зарплаты сотрудников на 1000 евро, выпуск возрастает на 0.02 млн евро при прочих равных условиях